textattack/search_methods/metropolis_hastings.py

```diff
@@ -0,0 +1,175 @@
+            torch.cuda.current_blas_handle()
+            self.language_model = self.language_model.to(utils.get_device())
+            self.use_gpu = True
+        except RuntimeError as error:
+            if "CUDA out of memory" in str(error):
+                warnings.warn("CUDA out of memory. Running GPT-2 for Metropolis Hastings on CPU.")
+                self.language_model = self.language_model.to("cpu")
+                self.use_gpu = False
+            else:
+                raise error
+        
+        self.language_model.eval()
+
+    def _lm_score(self, text):
+        """
+        Assigns likelihood of a text as 1/perplexity(text)
+        Args:
+            text (str)
+        Returns: 1/perplexity(text)
+        """
+
+        input_tokens = self.lm_tokenizer.tokenize(text, max_length=self.lm_tokenizer.model_max_length-2)
+        # Occasionally, len(input_tokens) != 1022, so we have to check it
+        if len(input_tokens) != self.lm_tokenizer.model_max_length-2:
+            input_tokens = input_tokens[:self.lm_tokenizer.model_max_length-2]
+        input_tokens.insert(0, self.lm_tokenizer.bos_token)
+        input_tokens.append(self.lm_tokenizer.eos_token_id)
+        input_ids = self.lm_tokenizer.convert_tokens_to_ids(input_tokens)
+        input_ids = torch.tensor(input_ids)
+        if self.use_gpu:
+            input_ids = input_ids.to(utils.get_device())
+        
+        with torch.no_grad():
+            loss = self.language_model(input_ids, labels=input_ids)[0].item()
+            del input_ids
+        
+        perplexity = math.exp(loss)
+        return 1/perplexity
+        
+    def _stationary_dist(self, x, original_label):
+        """
+        Unnormalized estimation of the distribution that we want to sample from.
+        Args:
+            x (TokenizedText)
+            original_label
+        Returns: lm_score(x) * prob_model(wrong|x)
+        """
+        model_result = self.get_goal_results([x], original_label)[0][0]
+        lm_score = self._lm_score(x.text)
+        return lm_score * model_result.score
+
+    def _batch_stationary_dist(self, x_list, original_label):
+        """
+        Unnormalized estimation of the distribution that we want to sample from.
+        Process items in batch.
+        Args:
+            x_list (list): List of TokenizedText
+            original_label
+        Returns: list of floats representing lm_score(x) * prob_model(wrong|x)
+        """
+        batch_size = len(x_list)
+        model_results = self.get_goal_results(x_list, original_label)[0]
+        lm_scores = [self._lm_score(x.text) for x in x_list]
+        scores = [lm_scores[i] * model_results[i].score for i in range(batch_size)]
+
+        return scores, model_results
+
+    def _normalize(self, values):
+        """
+        Take list of values and normalize it into a probability distribution
+        """
+        s = sum(values)
+        if s == 0:
```

Comment:
question: I think I understand what this is doing, but when do these ever sum to 0?

Reply:
Normally that never happens, but we have one or two cases where it does with LSTM-IMDB. It's because LSTM's confidence for original label is 1, so our `result.score` is `0`, so all the values are `0`.